[Feature/social-kakao]: 카카오 소셜 로그인 기능 구현

whswls opened the pull request:
## 작업한 내용
- Kakao SDK 초기화 코드 작성
- KakaoLoginViewModel 구현
- 카카오톡 앱/웹 로그인 처리
- 로그인 성공 시 토큰 저장 및 사용자 정보 처리
- 로그인 상태에 따라 메인 or 회원가입 화면 이동
- UserState에 소셜 로그인 응답 처리 로직 추가
- SocialLoginError enum 도입 및 상세 오류 케이스 분류
- ToastState를 활용한 오류 메시지 처리

## PR Point
- 앱 설치 여부에 따라 카카오톡 앱 로그인 / 웹 로그인 분기 처리
- JWT 토큰 키체인 저장 및 디버깅용 로그 출력 추가
    - AuthManager에 saveToken, debugPrintToken 메소드 작성 
- 강제 타입 캐스팅(as! T) 제거 및 제네릭 + 프로토콜을 활용한 안전한 타입 처리
- 소셜 로그인 실패 시 ToastState를 활용한 오류 메시지 처리

### 추후 작업할 내용 (미완성 부분)
<!-- 이슈 번호와 함께 향후 작업 예정인 내용을 작성해주세요 -->
<!-- 연결된 이슈가 있다면 작성해주세요 (예: #123) -->
<!-- 체크리스트 작성 (예: - [ ]) -->
- 서버 URL 설정 분리 및 환경에 따른 처리
    - 현재 'localhost'로 설정되어 있어 실제 기기에서 연결이 불가능

## 실행 영상
<!-- UI 변경이 있는 경우 변경 전/후 스크린샷을 첨부해주세요 -->
<table>
  <tr>
    <td>
      <img src="https://github.com/user-attachments/assets/3fc52e2c-7936-49ce-8dcd-c975daef486d" width="200"><br>
      <p>카카오 앱 연결 로그인</p>
    </td>
    <td>
      <img src="https://github.com/user-attachments/assets/b5c71611-c5b4-4515-8e96-c89110b6ee1f" width="200"><br>
      <p>카카오 웹 연결 로그인</p>
    </td>
  </tr>
</table>

whswls committed:
[Add]: 카카오 SDK 초기화 코드 작성

whswls committed:
[Add]: KakaoLoginView에 ViewModel 바인딩

whswls committed:
[Add]: 카카오 로그인 기능 구현

- 카카오톡 설치 여부에 따른 로그인 방식 분기
- 로그인 성공 시 accessToken으로 서버에 로그인 요청 코드 작성

whswls committed:
[Ref]: 소셜 로그인 응답 구조에 맞춰 User 모델 리팩토링

- message 속성을 Optional로 변경해 유연한 응답 처리 가능하도록 개선
- 신규 SocialLoginResponse 모델 추가
- UserData 모델에 서버 필드명과 대응하는 CodingKeys 명시적으로 지정

whswls committed:
[Add]: 응답 객체에 statusCode 자동 주입 기능 추가

whswls committed:
[Ref]: 소셜 로그인 응답 모델 SocialLoginResponse로 통일

- 소셜 로그인 함수의 반환 타입을 변경

whswls committed:
[Fix]: 사용자 정보 복원 및 토큰 저장 로직의 안정성 개선

- 응답 데이터 검증 시 토큰 존재 여부를 옵셔널 기반으로 안전하게 체크
- 토큰 저장 시 nil 가능성을 대비해 기본값 처리 추가

whswls committed:
[Add]: 소셜 로그인 응답 상태에 따라 화면 전환 처리

- 서버 응답 상태코드에 따라 화면(메인화면/회원가입 화면) 분기

whswls committed:
[Add]: 키체인 토큰 출력용 메소드 추가

- 디버깅 목적의 메소드 구현
- 키체인에 저장된 토큰 값 존재 여부 로그로 출력

whswls committed:
[Ref]: 로그인 흐름 정리 및 상태코드 기반 분기 로직 개선

- 서버 응답 후 토큰 저장 및 디버깅 로그 추가
- OAuth 오류 처리 및 토큰 유효성 검증 구조 정리

whswls committed:
[Ref]: 카카오 로그인 로직 중복 제거

- UserState 의존성 주입 추가 및 상태 기반 처리 정리

whswls committed:
[Fix]: 소셜 로그인 응답 검증 로직 개선

- token 검증 시 옵셔널 안전성 확보 및 공백 제거 후 isEmpty 확인

whswls committed:
[Ref]: 상태 코드 기반 화면 분기 로직 UserState로 이동

whswls committed:
[Ref]: 토큰 디버깅 로그 위치를 UserState로 이동하여 책임 분리

whswls committed:
[Ref]: 상태코드 분기 및 결과 반환 구조 개선

- 기존 회원, 신규 회원 상태에 따른 결과 반환을 위한 SocialLoginResult enum 추가

whswls committed:
[Ref]: 로그인 결과에 따른 화면 전환 로직 추가

- socialLogin 결과 분기 처리 추가

whswls committed:
[Ref]: 화면 전환 방식 변경

- fullScreenCover로 변경하여 화면 전환 방식 개선

unib35 commented:
코드 구조화, 오류 처리, 그리고 상태 관리 측면에서 정말 잘 작업해주셨네요! 몇 가지 개선할 수 있는 부분들을 아래에 정리했습니다. 큰 문제는 없는것 같아보이네요ㅎㅎ 대부분 코드 품질과 안정성을 더 높이기 위해 디테일 부분적인 제안드립니다.

unib35 commented on CineHive/CineHive/Features/Auth/View/SocialLogin/KakaoLoginBtnView.swift:
NavigationStack으로 Button을 감싸는 구조는 불필요하며 예상치 못한 동작을 유발할 수 있습니다. 아래와 같이 단순화하는 것이 좋겠습니다.
```swift
struct KakaoLoginBtnView: View {
    @State private var viewModel = KaKaoLoginViewModel()

    var body: some View {
        Button(action: {
            viewModel.login()
        }, label: {
            HStack {
                Image("KakaoLogo")
                    .resizable()
                    .frame(width: 18, height: 18)
                Text("카카오로 계속하기")
                    .font(.system(size: 15, weight: .bold))
                    .foregroundStyle(.black)
            }
            .frame(width: 330, height: 44)
            .background(Color("KakaoColor"))
            .clipShape(RoundedRectangle(cornerRadius: 12))
        })
        .fullScreenCover(isPresented: $viewModel.shouldNavigateToMain) {
            MainTabView()
        }
        .fullScreenCover(isPresented: $viewModel.shouldNavigateToSignUp) {
            SignUpView()
        }
    }
}
```
NavigationStack은 화면 전환을 관리하는 컨테이너로, 버튼 하나를 감싸는 용도로는 적합하지 않습니다. fullScreenCover만으로도 충분히 화면 전환이 가능합니다.

unib35 replied:
NavigationStack 사용이 필요한 경우는
- 계층적 화면 탐색이 필요할 때 (push/pop 네비게이션)
- 네비게이션 바(타이틀, 백 버튼 등)를 사용할 때
- navigationDestination을 통해 경로 기반 네비게이션을 구현할 때

이 버튼 뷰에서는 위의 기능들을 사용하지 않고, fullScreenCover만 사용하여 모달 형태로 화면을 표시하고 있습니다. 만약 NavigationStack을 사용할 특별한 이유가 없다면 (예: 미래에 네비게이션 기능을 추가할 예정이라든지), 굳이 필요하지 않을 수 있습니다.

whswls replied:
이전에 사용한 .navigationLink를 .fullScreenCover로 수정한 후 NavigationStack을 함께 수정하지 못했네요! 감사합니다 :)

unib35 commented on CineHive/CineHive/Core/Network/NetworkManager.swift:
ResponseWithStatusCode 프로토콜 구현에서 타입 캐스팅(as! T)을 사용하는 부분이 잠재적인 런타임 오류를 발생시킬 수 있습니다. 제네릭과 프로토콜 확장을 활용하여 더 안전하게 구현하는 것이 좋을것  같습니다.
```swift
protocol ResponseWithStatusCode {
    var statusCode: Int? { get set }
}

extension ResponseWithStatusCode {
    mutating func injectStatusCode(_ code: Int) {
        self.statusCode = code
    }
}

// 사용 예:
func decodeResponse<T: Decodable>(data: Data, response: HTTPURLResponse) throws -> T {
    let decoded = try JSONDecoder().decode(T.self, from: data)
    
    if var responseWithCode = decoded as? any ResponseWithStatusCode {
        responseWithCode.injectStatusCode(response.statusCode)
        // 여기서 안전하게 타입으로 변환
        if let typedResponse = responseWithCode as? T {
            return typedResponse
        }
    }
    
    return decoded
}
```
위처럼 하면 강제 타입 캐스팅을 피하고 안전하게 코드 작성가능합니당.

unib35 commented on CineHive/CineHive/Core/State/UserState.swift:
SocialLoginResult enum에서 실패 케이스에 더 구체적인 정보가 포함되어야 할것 같습니다.
```swift
enum SocialLoginResult {
    case successNavigateToMain
    case successNavigateToSignUp
    case failure(SocialLoginError)
}

enum SocialLoginError: Error {
    case networkError(String)
    case serverError(statusCode: Int, message: String)
    case invalidData(String)
    case unknownError(String)
    
    var message: String {
        switch self {
        case .networkError(let message): return "네트워크 오류: \(message)"
        case .serverError(let code, let message): return "서버 오류 (\(code)): \(message)"
        case .invalidData(let message): return "데이터 오류: \(message)"
        case .unknownError(let message): return "알 수 없는 오류: \(message)"
        }
    }
}
```
위는 예시인데, 오류 종류에 따라서 구체적인 처리 가능해지니 구체적이게 변경 해주삽쇼.

unib35 commented on CineHive/CineHive/Features/Auth/ViewModel/KakaoLoginViewModel.swift:
Task 내부의 오류 처리를 구현 필요 (현재는 단순히 로그만 찍힘)
```swift
let loginHandler: (OAuthToken?, Error?) -> Void = { [weak self] (oauthToken, error) in
    guard let self = self else { return }
    
    if let error = error {
        Logger.log(.error, category: Logger.auth, message: "카카오 로그인 실패: \(error.localizedDescription)")
        // UI에 오류 알림 표시
        return
    }
    
    guard let token = oauthToken else {
        Logger.log(.error, category: Logger.auth, message: "OAuth 토큰 없음")
        // UI에 오류 알림 표시
        return
    }
    
    Task {
        do {
            let result = await self.userState.socialLogin(provider: .kakao, token: token.accessToken)
            
            // 메인 스레드에서 UI 업데이트
            await MainActor.run {
                switch result {
                case .successNavigateToMain:
                    self.shouldNavigateToMain = true
                case .successNavigateToSignUp:
                    self.shouldNavigateToSignUp = true
                case .failure(let message):
                    // 오류 처리 및 UI에 표시
                    print("로그인 실패:", message)
                }
            }
        } catch {
            await MainActor.run {
                // Task 자체의 오류 처리
                print("비동기 처리 중 오류 발생:", error.localizedDescription)
            }
        }
    }
}
```
위 예시 참고해서 오류 처리 추가 해주삽쇼
구현되어 있는 토스트 써도되고 다른 아이디어 있으면 구현해보는것도 나쁘지 않을듯합니다.

unib35 reviewed (commented):
PR 설명에 언급된 대로 서버 URL을 환경(개발/스테이징/프로덕션)에 따라 설정할 수 있는 구조를 추가하면 좋을것 같습니다. 

```swift
enum Environment {
    case development
    case staging
    case production
    
    var baseURL: URL {
        switch self {
        case .development:
            return URL(string: "http://localhost:8080")!
        case .staging:
            return URL(string: "https://staging-api.example.com")!
        case .production:
            return URL(string: "https://api.example.com")!
        }
    }
}

// 사용 예:
#if DEBUG
let environment: Environment = .development
#else
let environment: Environment = .production
#endif

let baseURL = environment.baseURL
```
이렇게 하면 각 환경별로 서로 다른 URL을 교체해서 사용할 수 있음

whswls committed:
[Ref]: .fullScreenCover 도입에 따라 NavigationStack 제거

whswls committed:
[Ref]: 서버 환경별 baseURL 설정 구조화

- 손쉽게 서버 주소 전환 가능
- enum 추가로 환경(개발, 스테이징, 프로덕션) 구분

whswls committed:
[Ref]: 상태 코드 주입용 ResponseWithStatusCode 프로토콜 분리 및 타입 캐스팅 개선

- 기존 강제 타입 캐스팅(as! T) 방식 제거
- 제네릭 타입과 프로토콜 기반으로 안전하게 리팩토링

whswls committed:
[Ref]: 소셜 로그인 오류 타입을 구체적인 에러 케이스로 확장

- SocialLoginError enum에 네트워크 오류, 시스템- - 클라이언트 오류, 디코딩 오류 등 다양한 상황에 대응하는 세부 케이스 추가
- 각 케이스에 맞는 메시지를 제공
- 에러 발생 시 더 명확하고 구체적인 피드백이 가능하도록 구조 개선

whswls committed:
[Ref]: KakaoLoginViewModel내 ToastState를 활용한 오류 메시지 표시 로직 추가

- 로그인 실패 시 사용자에게 오류 메시지를 표시

unib35 commented:
@whswls님! 코드 리뷰 피드백 반영해주셔서 감사합니다. 변경 사항들 살펴봤는데, 정말 훌륭하게 개선되었네요! 👏
